driver-core/src/main/com/mongodb/internal/logging/LogMessage.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.logging;
+
+import com.mongodb.connection.ClusterId;
+import com.mongodb.lang.Nullable;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public final class LogMessage {
```

Comment:
I renamed the StructuredLogMessage class to LogMessage. The reason for this change is that by design and specification, StructuredLogMessage only allows non-null entry values. However, we also need to preserve null values when interpolating a message to indicate which parts should be omitted.

To capture all entries, both null and non-null, new wrapper class (LogMessage) was introduced that allows to contain all entries with their corresponding values. Additionally, StructuredLogMessage is an inner class of LogMessage, so that LogMessage can be converted to StructuredLogMessage using the `toStructuredLogMessage `method, which creates a null-free structured log  message.

To handle message interpolation,  UnstructuredLogMessage inner class class is introduced as well, which triggers interpolation logic lazily. Our Logger captures LogMessage and internally converts it to either an Unstructured or Structured log message. At this time, we only emit unstructured logs to the user.